cpp/skyweaver/detail/MultiFileWriter.cu

```diff
@@ -241,23 +242,73 @@ bool MultiFileWriter<VectorType>::operator()(VectorType const& stream_data,
                                              std::size_t stream_idx)
 {
     if(!has_stream(stream_idx)) {
+        if (_config.get_wait_config().is_enabled) {
+            try {
+                wait_for_space();
+            } catch (std::runtime_error& e) {
+                std::cout << "Wait loop exception: " << e.what() << std::endl;
+                throw;
+            }
+        }
+
         create_stream(stream_data, stream_idx);
     }
     if constexpr(is_device_vector<VectorType>::value) {
         thrust::host_vector<typename VectorType::value_type> stream_data_h =
             stream_data.vector();
+        if (_config.get_wait_config().is_enabled) {
+            try {
+                wait_for_space();
+            } catch (std::runtime_error& e) {
+                std::cout << "Wait loop exception: " << e.what() << std::endl;
+                throw;
+            }
+        }
         _file_streams.at(stream_idx)
             ->write(reinterpret_cast<char const*>(stream_data_h.data()),
-                    stream_data_h.size() *
-                        sizeof(typename VectorType::value_type));
+                stream_data_h.size() * sizeof(typename VectorType::value_type));
     } else {
+        if (_config.get_wait_config().is_enabled) {
+             try {
+                 wait_for_space();
+             } catch (std::runtime_error& e) {
+                 std::cout << "Wait loop exception: " << e.what() << std::endl;
+                 throw;
+             }
+        }
         _file_streams.at(stream_idx)
-            ->write(reinterpret_cast<char const*>(
-                        thrust::raw_pointer_cast(stream_data.data())),
-                    stream_data.size() *
-                        sizeof(typename VectorType::value_type));
+            ->write(reinterpret_cast<char const*>( thrust::raw_pointer_cast(stream_data.data())),
+                stream_data.size() * sizeof(typename VectorType::value_type));
     }
     return false;
 }
 
-} // namespace skyweaver
\ No newline at end of file
+template <typename VectorType>
+void MultiFileWriter<VectorType>::wait_for_space()
+{
+    std::filesystem::space_info space = std::filesystem::space(_config.output_dir());
+    if(space.available >= _config.get_wait_config().min_free_space)
```

Comment:
I think this should be tied to the size of the write you want to do. e.g. it should be a percentage overhead of the write size.

Reply:
It is a design decision. Currently, it is solely configurable. There might be potential race condition with other programs writing on the same disk. If this doesn't happen, we can compute the size of the write. Should the overhead configurable then. or will we give it just a factor of two?

Reply:
50% extra than the write sounds fine. The worry of leaving it a fixed size is that you could end up with a write that is larger than the min_free_space.

Another, maybe preferable, way to do this is to specify that the min_free_space is the required free space AFTER the write, i.e. it is specifying some safety buffer zone of RAM that nominally should never be filled, but might be if there are race conditions (but if large enough this should not matter).